src/Client.cpp

```diff
@@ -1,245 +1,487 @@
     uint8_t crc = uint8_t(data.size())^id;
     for(const uint8_t d : data) { crc = crc^d; }
     return crc;
 }
 
-void Client::processOneMessage() {
-    uint8_t c;
-    // find start of header
-    while(true) {
-        // find '$'
-        for(c=0; c!='$'; asio::read(pimpl->port, asio::buffer(&c,1)));
-        // check 'M'
-        asio::read(pimpl->port, asio::buffer(&c,1));
-        if(c=='M') { break; }
+ByteVector Client::packMessageV2(const msp::ID id, const ByteVector &data) 
+{
+    
+    ByteVector msg;
+    msg.push_back('$');                                 // preamble1
+    msg.push_back('X');                                 // preamble2
+    msg.push_back('<');                                 // direction
+    msg.push_back(0);                                   // flag
+    msg.push_back(uint8_t( uint16_t(id) & 0xFF ));                // message_id low bits
+    msg.push_back(uint8_t( uint16_t(id) >> 8 ));                  // message_id high bits
+    
+    uint16_t size = (uint16_t)data.size();
+    msg.push_back(uint8_t( size & 0xFF ));              // data size low bits
+    msg.push_back(uint8_t( size >> 8 ));                // data size high bits
+    
+    msg.insert(msg.end(), data.begin(), data.end());    // data
+    msg.push_back( crcV2( 0, ByteVector(msg.begin()+3,msg.end()) ) );                     // crc
+    
+    return msg;
+
+}
+
+uint8_t Client::crcV2(uint8_t crc, const ByteVector &data) 
+{
+    for (const uint8_t& p : data) {
+        crc = crcV2(crc,p);
     }
+    return crc;
+}
 
-    // message direction
-    asio::read(pimpl->port, asio::buffer(&c,1));
-    const bool ok_id = (c!='!');
 
+uint8_t Client::crcV2(uint8_t crc, const uint8_t& b) 
+{
+    crc ^= b;
+    for (int ii = 0; ii < 8; ++ii) {
+        if (crc & 0x80) {
+            crc = (crc << 1) ^ 0xD5;
+        } else {
+            crc = crc << 1;
+        }
+    }
+    return crc;
+}
+
+
+
+void Client::processOneMessage(const asio::error_code& ec,std::size_t bytes_transferred) {
+    if (print_debug_) std::cout << "processOneMessage on " << bytes_transferred << " bytes" << std::endl;
+
+    if (ec == asio::error::operation_aborted) {
+        //operation_aborted error probably means the client is being closed
+        // notify waiting request methods
+        cv_response.notify_all();
+        return;
+    } 
+    
+    // ignore and remove header bytes
+    //buffer.consume(bytes_transferred);
+    const uint8_t msg_marker = extractChar();
+    if (msg_marker != '$') std::cerr << "Message marker " << msg_marker << " is not recognised!" << std::endl;
+    
+    // message version
+    int ver = 0;
+    const uint8_t ver_marker = extractChar();
+    if (ver_marker == 'M') ver = 1;
+    if (ver_marker == 'X') ver = 2;
+    if (ver == 0) {
+        std::cerr << "Version marker " << ver_marker << " is not recognised!" << std::endl;
+    }
+    
+    ReceivedMessage recv_msg;
+    if (ver == 2)
```

Comment:
Do we need to detect the MSP version here? The FC should respond with the protocol format that was also used to send messages. That is, the received MSP version should be the same as the one currently set in the `Client`. Hence, we only need to check if the received version is the correct one, and continue with the version set in `Client`.

Reply:
Do we need to do the MSP version detection here or can we just expect that the FC will respond with the same protocol version at which we sent messages to the FC?